app/scripts/controllers/transactions/tx-state-manager.js

```diff
@@ -244,255 +244,328 @@ export default class TransactionStateManager extends EventEmitter {
-    from: '0x0..',<br>
-    status: 'signed', \\ (status) => status !== 'rejected' give me all txs who's status is not rejected<br>
-    err: undefined,<br>
-  }<br></code>
-  optionally the values of the keys can be functions for situations like where
-  you want all but one status.
-  @param {Array} [initialList=this.getTxList()]
-  @returns {Array} array of txMeta with all
-  options matching
-  */
-  /*
-  ****************HINT****************
-  | `err: undefined` is like looking |
-  | for a tx with no err             |
-  | so you can also search txs that  |
-  | dont have something as well by   |
-  | setting the value as undefined   |
-  ************************************
-
-  this is for things like filtering a the tx list
-  for only tx's from 1 account
-  or for filtering for all txs from one account
-  and that have been 'confirmed'
-  */
-  getFilteredTxList(opts, initialList) {
-    let filteredTxList = initialList;
-    Object.keys(opts).forEach((key) => {
-      filteredTxList = this.getTxsByMetaData(key, opts[key], filteredTxList);
+  getTransactions({
+    searchCriteria = {},
+    initialList,
+    filterToCurrentNetwork = true,
+    limit,
+  } = {}) {
+    const chainId = this.getCurrentChainId();
+    const network = this.getNetwork();
+    // searchCriteria is an object that might have values that aren't predicate
+    // methods. When providing any other value type (string, number, etc), we
+    // consider this shorthand for "check the value at key for strict equality
+    // with the provided value". To conform this object to be only methods, we
+    // mapValues (lodash) such that every value on the object is a method that
+    // returns a boolean.
+    const predicateMethods = mapValues(searchCriteria, (predicate) => {
+      return typeof predicate === 'function'
+        ? predicate
+        : (v) => v === predicate;
     });
-    return filteredTxList;
-  }
 
-  /**
-   * @param {string} key - the key to check
-   * @param {any} value - the value your looking for can also be a function that returns a bool
-   * @param {Array} [txList=this.getTxList()] - the list to search. default is the txList
-   *  from txStateManager#getTxList
-   * @returns {Array} a list of txMetas who matches the search params
-   */
-  getTxsByMetaData(key, value, txList = this.getTxList()) {
-    const filter = typeof value === 'function' ? value : (v) => v === value;
+    // If an initial list is provided we need to change it back into an object
+    // first, so that it matches the shape of our state. This is done by the
+    // lodash keyBy method. This is the edge case for this method, typically
+    // initialList will be undefined.
+    const transactionsToFilter = initialList
+      ? keyBy(initialList, 'id')
+      : this.store.getState().transactions;
+
+    // Combine sortBy and pickBy to transform our state object into an array of
+    // matching transactions that are sorted by time.
+    const filteredTransactions = sortBy(
```

Comment:
The previous iteration of this logic resulted in the potential for the entire transaction list to be iterated through N + 1 times where N was the number of search terms. The + 1 is because the entire list was iterated through once to get only the current network transactions.

This code should both be more performant and still return transactions in the order in which they were expected prior to my changes.